`assert_type()` says an int is not an int

The following code:

```python
from typing_extensions import assert_type

assert_type(42, int)
```

fails to typecheck with the following message:

```
assert-type01.py:3: error: Expression is of type "int", not "int"
```

**Your Environment**

- Mypy version used: both 0.960 and commit 1636a0549670e1b75e2987c16ef26ebf91dfbde9
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.9.13
- Operating system and version: macOS 11.6.6

```
(venv) carl@x1:~/src/mypy$ python3 runtests.py self
run self: ['/home/carl/src/mypy/venv/bin/python3', '-m', 'mypy', '--config-file', 'mypy_self_check.ini', '-p', 'mypy']
mypy/checkexpr.py:3924: error: Never apply isinstance() to unexpanded types; use mypy.types.get_proper_type() first  [misc]
            if isinstance( source_type, mypy.types.Instance ) and not is_same_type(source_type.last_known_value, target_t...
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
mypy/checkexpr.py:3924: note: If you pass on the original type after the check, always use its unexpanded version
mypy/checkexpr.py:3924: error: Argument 1 to "is_same_type" has incompatible type "Optional[LiteralType]"; expected "Type" 
[arg-type]
            if isinstance( source_type, mypy.types.Instance ) and not is_same_type(source_type.last_known_value, target_t...
                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
mypy/checkexpr.py:3930: error: Argument 1 to "assert_type_fail" of "MessageBuilder" has incompatible type
"Optional[LiteralType]"; expected "Type"  [arg-type]
                self.msg.assert_type_fail(source_type.last_known_value, target_type, expr)
                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
Found 3 errors in 1 file (checked 170 source files)
```
working on it...

```
(venv) carl@x1:~/src/ck/mypy$ python3 runtests.py self 
run self: ['/home/carl/src/mypy/venv/bin/python3', '-m', 'mypy', '--config-file', 'mypy_self_check.ini', '-p', 'mypy']
mypy/checkexpr.py:3923: error: "Type" has no attribute "last_known_value"  [attr-defined]
                    and source_type.last_known_value is not None:
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
`+                and (hasattr(source_type, "last_known_value") and source_type.last_known_value is not None):`

Success: no issues found in 170 source files

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -3918,6 +3918,9 @@ class ExpressionChecker(ExpressionVisitor[Type]):
             always_allow_any=True,
         )
         target_type = expr.type
+        if isinstance( source_type, mypy.types.Instance ) \
+                and source_type.last_known_value is not None:
+            source_type = source_type.last_known_value
         if not is_same_type(source_type, target_type):
             if not self.chk.in_checked_function():
                 self.msg.note(

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -3918,9 +3918,8 @@ class ExpressionChecker(ExpressionVisitor[Type]):
             always_allow_any=True,
         )
         target_type = expr.type
-        if (
-                isinstance( mypy.types.get_proper_type(source_type), mypy.types.Instance ) \
-                and (hasattr(source_type, "last_known_value") and source_type.last_known_value is not None)
+        if isinstance(mypy.types.get_proper_type(source_type), mypy.types.Instance) and (
+            hasattr(source_type, "last_known_value") and source_type.last_known_value is not None
         ):
             source_type = source_type.last_known_value
         if not is_same_type(source_type, target_type):

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -3918,10 +3918,9 @@ class ExpressionChecker(ExpressionVisitor[Type]):
             always_allow_any=True,
         )
         target_type = expr.type
-        if isinstance(mypy.types.get_proper_type(source_type), mypy.types.Instance) and (
-            hasattr(source_type, "last_known_value") and source_type.last_known_value is not None
-        ):
-            source_type = source_type.last_known_value
+        proper_source_type = get_proper_type(source_type)
+        if isinstance(proper_source_type, mypy.types.Instance) and proper_source_type.last_known_value is not None:
+            source_type = proper_source_type.last_known_value
         if not is_same_type(source_type, target_type):
             if not self.chk.in_checked_function():
                 self.msg.note(

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -3919,7 +3919,10 @@ class ExpressionChecker(ExpressionVisitor[Type]):
         )
         target_type = expr.type
         proper_source_type = get_proper_type(source_type)
-        if isinstance(proper_source_type, mypy.types.Instance) and proper_source_type.last_known_value is not None:
+        if (
+            isinstance(proper_source_type, mypy.types.Instance)
+            and proper_source_type.last_known_value is not None
+        ):
             source_type = proper_source_type.last_known_value
         if not is_same_type(source_type, target_type):
             if not self.chk.in_checked_function():

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -3918,8 +3918,8 @@ class ExpressionChecker(ExpressionVisitor[Type]):
             always_allow_any=True,
         )
         target_type = expr.type
-        if isinstance( source_type, mypy.types.Instance ) \
-                and source_type.last_known_value is not None:
+        if isinstance( mypy.types.get_proper_type(source_type), mypy.types.Instance ) \
+                and (hasattr(source_type, "last_known_value") and source_type.last_known_value is not None):
             source_type = source_type.last_known_value
         if not is_same_type(source_type, target_type):
             if not self.chk.in_checked_function():

--- a/test-data/unit/check-expressions.test
+++ b/test-data/unit/check-expressions.test
@@ -937,6 +937,8 @@ reveal_type(returned)  # N: Revealed type is "builtins.int"
 assert_type(a, str)  # E: Expression is of type "int", not "str"
 assert_type(a, Any)  # E: Expression is of type "int", not "Any"
 assert_type(a, Literal[1])  # E: Expression is of type "int", not "Literal[1]"
+assert_type(42, Literal[42])
+assert_type(42, int)  # E: Expression is of type "Literal[42]", not "int"
 [builtins fixtures/tuple.pyi]
 
 [case testAssertTypeGeneric]

